src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -315,15 +316,124 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
                                                            chip::ByteSpan ICACertificate, chip::ByteSpan IPKValue,
                                                            chip::NodeId CaseAdminNode, uint16_t AdminVendorId)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_FAILURE;
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> cert;
+
+    uint8_t * certBuf = nullptr;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has added an Op Cert");
+
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    // Fetch current admin
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    VerifyOrExit(cert.Alloc(NOC.size() + ICACertificate.size()), status = EMBER_ZCL_STATUS_FAILURE);
+    certBuf = cert.Get();
+    memmove(certBuf, NOC.data(), NOC.size());
```

Comment:
why memmove? do we expect overlapping buffers? memcpy seems sufficient as certBuf is the allocated cert.

I believe there are other places where I see memmove. Could we do memcpy for all? rationale is that whenever I see memmove I start to think on why buffers overlap.

Reply:
++

It's a good point.  Unnecessary memmove creates a similar maintenance problem to over-provisioned locking: future maintainers will be terrified to change it to memcpy for fear of not understanding some subtle reason why the buffers could overlap, just as future maintainers might be terrified to remove a mutex lock for fear of not seeing some subtle reason why locking is needed.

In this case, we know the buffers don't overlap, so we should use memcpy.